File: headless/src/main/org/nlogo/prim/_patch.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.api.LogoException;
import org.nlogo.api.Nobody$;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Reporter;

public final strictfp class _patch
    extends Reporter {
  @Override
  public Syntax syntax() {
    int[] right;
    right = new int[]{Syntax.NumberType(),
        Syntax.NumberType()};

    return Syntax.reporterSyntax
        (right, Syntax.PatchType() | Syntax.NobodyType());
  }

  // I've tried to rejigger this and the result gets past TryCatchSafeChecker but then
  // doesn't work at runtime ("Inconsistent stack height") - ST 2/10/09
  @Override
  public Object report(final Context context) throws LogoException {
    try {
      return
          world.getPatchAt
              (argEvalDoubleValue(context, 0),
                  argEvalDoubleValue(context, 1));
    } catch (org.nlogo.api.AgentException ex) {
      return Nobody$.MODULE$;
    }
  }
}

File: src/main/prim/etc/_patch.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc

import org.nlogo.api.{ Syntax, AgentException, Nobody }
import org.nlogo.nvm.{ Reporter, Context }

class _patch extends Reporter {

  override def syntax =
    Syntax.reporterSyntax(
      Array(Syntax.NumberType, Syntax.NumberType),
      Syntax.PatchType | Syntax.NobodyType)

  // I've tried to rejigger this and the result gets past TryCatchSafeChecker but then doesn't work
  // at runtime ("Inconsistent stack height") - ST 2/10/09
  override def report(context: Context): AnyRef =
    try world.getPatchAt(
      argEvalDoubleValue(context, 0),
      argEvalDoubleValue(context, 1))
    catch {
      case _: AgentException => Nobody
    }

}
